### PR TITLE
[Console] Add types to private properties

### PR DESCRIPTION
**Pull request by derrabus**
| Q             | A
| ------------- | ---
| Branch?       | 6.0
| Bug fix?      | no
| New feature?  | no
| Deprecations? | no
| Tickets       | N/A
| License       | MIT
| Doc PR        | N/A

### CODE REVIEW

File: src/Symfony/Component/Console/Output/StreamOutput.php [hunk 1]
```diff
@@ -29,7 +29,10 @@
  */
 class StreamOutput extends Output
 {
-    private $stream;
+    /**
+     * @var resource
+     */
```
Comment (chalasr) [inline src/Symfony/Component/Console/Output/StreamOutput.php]: why is it needed?

**Review by chalasr — COMMENTED** [review]
rebase needed

**Comment by nicolas-grekas** [comment]
(rebase needed)

File: src/Symfony/Component/Console/Application.php [hunk 2]
```diff
@@ -250,12 +250,12 @@ public function doRun(InputInterface $input, OutputInterface $output)
...
 
         try {
-            $this->runningCommand = null;
+            unset($this->runningCommand);
             // the command name MUST be the first element of the input
             $command = $this->find($name);
         } catch (\Throwable $e) {
             if (!($e instanceof CommandNotFoundException && !$e instanceof NamespaceNotFoundException) || 1 !== \count($alternatives = $e->getAlternatives()) || !$input->isInteractive()) {
-                if (null !== $this->dispatcher) {
+                if (isset($this->dispatcher)) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Console/Application.php]: why is this better than making the property nullable?
making the property not-nullable might hide the fact that this check is going to be needed before using the propery. Making it nullable would make that explicit, isn't it?

File: src/Symfony/Component/Console/Command/Command.php [hunk 3]
```diff
@@ -394,11 +396,8 @@ public function getDefinition(): InputDefinition
      */
     public function getNativeDefinition(): InputDefinition
     {
-        if (null === $this->definition) {
-            throw new LogicException(sprintf('Command class "%s" is not correctly initialized. You probably forgot to call the parent constructor.', static::class));
-        }
-
-        return $this->definition;
+        return $this->definition
+            ?? throw new LogicException(sprintf('Command class "%s" is not correctly initialized. You probably forgot to call the parent constructor.', static::class));
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Console/Command/Command.php]: one line please:)

File: src/Symfony/Component/Console/Event/ConsoleEvent.php [hunk 4]
```diff
@@ -25,10 +25,10 @@ class ConsoleEvent extends Event
 {
     protected $command;
 
-    private $input;
-    private $output;
+    private InputInterface $input;
+    private OutputInterface $output;
 
-    public function __construct(Command $command = null, InputInterface $input, OutputInterface $output)
+    public function __construct(?Command $command, InputInterface $input, OutputInterface $output)
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Console/Event/ConsoleEvent.php:31]: for 4.4?

File: src/Symfony/Component/Console/Helper/ProgressBar.php [hunk 5]
```diff
@@ -413,8 +402,8 @@ public function display(): void
             return;
         }
 
-        if (null === $this->format) {
-            $this->setRealFormat($this->internalFormat ?: $this->determineBestFormat());
+        if (!isset($this->format)) {
+            $this->setRealFormat($this->internalFormat ?? $this->determineBestFormat());
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Console/Helper/ProgressBar.php]: potentially breaking change

File: src/Symfony/Component/Console/Output/StreamOutput.php [hunk 6]
```diff
@@ -29,7 +29,10 @@
...
+    private mixed $stream;
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Console/Output/StreamOutput.php]: better remove mixed if it's a resource

File: src/Symfony/Component/Console/Question/Question.php [hunk 7]
```diff
@@ -220,11 +220,8 @@ public function getValidator(): ?callable
      */
     public function setMaxAttempts(?int $attempts): static
     {
-        if (null !== $attempts) {
-            $attempts = (int) $attempts;
-            if ($attempts < 1) {
-                throw new InvalidArgumentException('Maximum number of attempts must be a positive value.');
-            }
+        if (null !== $attempts && $attempts < 1) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Console/Question/Question.php]: for 5.4

File: src/Symfony/Component/Console/Tester/TesterTrait.php [hunk 8]
```diff
@@ -95,11 +96,8 @@ public function getOutput(): OutputInterface
      */
     public function getStatusCode(): int
     {
-        if (null === $this->statusCode) {
-            throw new \RuntimeException('Status code not initialized, did you execute the command before requesting the status code?');
-        }
-
-        return $this->statusCode;
+        return $this->statusCode
+            ?? throw new \RuntimeException('Status code not initialized, did you execute the command before requesting the status code?');
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Console/Tester/TesterTrait.php]: one line

File: src/Symfony/Component/Console/Output/StreamOutput.php (hunk 6 above)
Comment (derrabus) [reply src/Symfony/Component/Console/Output/StreamOutput.php]: But I won't achieve 100% typed private properties then! 😓

File: src/Symfony/Component/Console/Application.php (hunk 2 above)
Comment (derrabus) [reply src/Symfony/Component/Console/Application.php]: In situations like this, where we always have to safeguard the access anyway, we can probably go with a nullable property.

File: src/Symfony/Component/Console/Helper/ProgressBar.php (hunk 5 above)
Comment (derrabus) [reply src/Symfony/Component/Console/Helper/ProgressBar.php]: reverted

File: src/Symfony/Component/Console/Event/ConsoleEvent.php (hunk 4 above)
Comment (derrabus) [reply src/Symfony/Component/Console/Event/ConsoleEvent.php:31]: #42738

**Review by nicolas-grekas — COMMENTED** [review]
Do you think we can reduce the delta with 5.4?
I fear this might make future merges complex :(

**Comment by derrabus** [comment]
I'll try.

**Comment by nicolas-grekas** [comment]
Thank you. I know this is tedious work, sorry to make it a bit more... :)

File: src/Symfony/Component/Console/Helper/ProgressBar.php [hunk 9]
```diff
@@ -240,11 +232,8 @@ public function setBarCharacter(string $char)
 
     public function getBarCharacter(): string
     {
-        if (null === $this->barChar) {
-            return $this->max ? '=' : $this->emptyBarChar;
-        }
-
-        return $this->barChar;
+        return $this->barChar
+            ?? ($this->max ? '=' : $this->emptyBarChar);
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Console/Helper/ProgressBar.php]: let's put all those on one line please

**Comment by derrabus** [comment]
See #42743 for some backports. That should reduce the diff a bit.

**Comment by nicolas-grekas** [comment]
(rebase needed)

**Comment by derrabus** [comment]
Rebased. The diff looks much better now, I think.

**Comment by nicolas-grekas** [comment]
Thank you @derrabus.

**Pull request merged**